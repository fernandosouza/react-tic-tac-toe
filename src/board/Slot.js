import React from 'react';
import styled, { css, keyframes } from 'styled-components';

import { ReactComponent as X } from './x.svg';
import { ReactComponent as Circle } from './circle.svg';

const scale = keyframes`
  0% {
    transform: scale(1);
  }

  100% {
    transform: scale(1.1);
  }
`;

const Player1 = styled(X)`
  width: 100px;
  height: 100px;

  ${props => props.winner && css`
    animation: ${scale} infinite alternate ease-in-out .54s;
  `}
`

const Player2 = styled(Circle)`
  width: 100px;
  height: 100px;

  ${props => props.winner && css`
    animation: ${scale} infinite alternate ease-in-out .54s;
  `}
`

const SlotWrapper = styled.div`
  --dimensions: calc(var(--board-size) / 3);

  /*TODO: How to move this block to a shared place? */
  display: flex;
  align-items: center;
  justify-content: center;

  background-color: var(--base-color);
  border: var(--grid-border);
  height: var(--dimensions);
  width: var(--dimensions);
  transition: background-color 150ms linear;
  cursor: pointer;

  ${props => !props.player && css`
    :hover {
      background-color: #3e5368;
    }
  `}
`;

export const Slot = props => {
  return (
    <SlotWrapper
      data-testid={`Slot${props.index}`}
      onClick={() => props.onSlotClick(props.index)}
    >
      {
        {
          1: <Player1 winner={props.winner} />,
          2: <Player2 winner={props.winner} />
        }[props.player]
      }
    </SlotWrapper>
  );
};
