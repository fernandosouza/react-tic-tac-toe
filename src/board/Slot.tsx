import React, { useContext, useState, FC, useEffect } from 'react';
import styled, { css, keyframes } from 'styled-components';

import { ReactComponent as X } from './x.svg';
import { ReactComponent as Circle } from './circle.svg';
import { GameContext } from '../GameContext';
import { WinnerSlots } from '../ticTacToe/TicTacToe';

const scale = keyframes`
  0% {
    transform: scale(1);
  }

  100% {
    transform: scale(1.1);
  }
`;

const Player1 = styled(X)<{ winner?: boolean }>`
  width: 100px;
  height: 100px;

  ${props => props.winner && css`
    animation: ${scale} infinite alternate ease-in-out .54s;
  `}
`

const Player2 = styled(Circle)<{ winner?: boolean }>`
  width: 100px;
  height: 100px;

  ${props => props.winner && css`
    animation: ${scale} infinite alternate ease-in-out .54s;
  `}
`

const SlotWrapper = styled.button<{ player: boolean }>`
  --dimensions: calc(var(--board-size) / 3);

  /*TODO: How to move this block to a shared place? */
  display: flex;
  align-items: center;
  justify-content: center;

  background-color: var(--base-color);
  border: var(--grid-border);
  height: var(--dimensions);
  width: var(--dimensions);
  transition: background-color 150ms linear;
  cursor: pointer;

  ${props => !props.player && css`
    :hover {
      background-color: #3e5368;
    }
  `}
`;

export const Slot: FC<{ index: number }> = props => {
  const [player, setPlayer] = useState<null | number>(null);
  const [winner, setWinner] = useState<null | WinnerSlots>(null);
  const gameContext = useContext(GameContext);

  const onSlotClick = (index: number) => {
    gameContext.game!.fillSlot(index);
    if (gameContext.game!.getBoard().get(index)) {
      setPlayer(gameContext.game!.getBoard().get(index));
    }
  };

  useEffect(() => {
    gameContext.game!.on('gameEnd', setWinner);
    return () => {
      gameContext.game!.off('gameEnd', setWinner);
    }
  }, []);

  return (
    <SlotWrapper
      disabled={!!winner}
      type="button"
      player={!!player}
      data-testid={`Slot${props.index}`}
      onClick={() => onSlotClick(props.index)}
    >
      {
        {
          1: <Player1 winner={winner && winner!.slots.includes(props.index) || undefined} />,
          2: <Player2 winner={winner && winner!.slots.includes(props.index) || undefined} />
          //@ts-ignore
        }[player]
      }
    </SlotWrapper>
  );
};
