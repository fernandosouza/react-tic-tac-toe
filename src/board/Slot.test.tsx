import React from 'react';
import { Slot } from './Slot';
import { create } from 'react-test-renderer';

describe('Slot', () => {
  it('renders slot', () => {
    const slot = create(<Slot index={1} />).toJSON();
    expect(slot).toMatchSnapshot();
  });
});
