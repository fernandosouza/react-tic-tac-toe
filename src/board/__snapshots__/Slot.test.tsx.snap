// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot renders slot 1`] = `
<button
  className="sc-htpNat elABiC"
  data-testid="Slot1"
  onClick={[Function]}
  type="button"
/>
`;
